File: packages/flint-py/flint/vegalite/templates/scatter.py
"""Scatter Plot, Regression, Ranged Dot Plot, and Boxplot templates."""
from __future__ import annotations

import copy

from .utils import (
    default_build_encodings,
    apply_point_size_scaling,
    set_mark_prop,
    detect_banded_axis_force_discrete,
)


def _scatter_instantiate(spec, ctx):
    default_build_encodings(spec, ctx["resolvedEncodings"])
    # A `shape` encoding only renders distinct glyphs on the `point` mark;
    # `circle` ignores it. Promote the mark when shape is in play.
    if (spec.get("encoding") or {}).get("shape", {}).get("field"):
        spec["mark"] = set_mark_prop(spec.get("mark"), "type", "point")
    canvas = ctx.get("canvasSize") or {}
    apply_point_size_scaling(spec, ctx["table"], canvas.get("width"), canvas.get("height"))
    config = ctx.get("chartProperties")
    if config and config.get("opacity") is not None and config["opacity"] < 1:
        spec["mark"] = set_mark_prop(spec.get("mark"), "opacity", config["opacity"])


scatter_plot_def = {
    "chart": "Scatter Plot",
    "template": {"mark": "circle", "encoding": {}},
    "channels": ["x", "y", "color", "size", "shape", "opacity", "column", "row"],
    "markCognitiveChannel": "position",
    "instantiate": _scatter_instantiate,
    "properties": [
        {"key": "opacity", "label": "Opacity", "type": "continuous",
         "min": 0.1, "max": 1, "step": 0.05, "defaultValue": 1},
    ],
}


# ─── Regression ─────────────────────────────────────────────────────────────

def _regression_instantiate(spec, ctx):
    encs = ctx["resolvedEncodings"]
    x = encs.get("x")
    y = encs.get("y")
    color = encs.get("color")
    size = encs.get("size")
    column = encs.get("column")
    row = encs.get("row")
    config = ctx.get("chartProperties")

    layer0 = spec["layer"][0]
    layer1 = spec["layer"][1]

    if x:
        layer0["encoding"]["x"] = {**layer0["encoding"].get("x", {}), **x}
        layer1["encoding"]["x"] = {**layer1["encoding"].get("x", {}), **x}
        if x.get("field"):
            layer1["transform"][0]["on"] = x["field"]
    if y:
        layer0["encoding"]["y"] = {**layer0["encoding"].get("y", {}), **y}
        layer1["encoding"]["y"] = {**layer1["encoding"].get("y", {}), **y}
        if y.get("field"):
            layer1["transform"][0]["regression"] = y["field"]

    method = (config or {}).get("regressionMethod")
    if method and method != "linear":
        layer1["transform"][0]["method"] = method
        if method == "poly":
            order = (config or {}).get("polyOrder", 3)
            layer1["transform"][0]["order"] = order

    if color:
        layer0["encoding"]["color"] = {**layer0["encoding"].get("color", {}), **color}
        if color.get("field"):
            layer1["transform"][0]["groupby"] = [color["field"]]
            layer1["encoding"]["color"] = {**color}
            layer1["mark"] = {"type": "line"}

    if size:
        layer0["encoding"]["size"] = {**layer0["encoding"].get("size", {}), **size}

    if "encoding" not in spec:
        spec["encoding"] = {}
    if column:
        spec["encoding"]["column"] = column
    if row:
        spec["encoding"]["row"] = row


regression_def = {
    "chart": "Regression",
    "template": {
        "layer": [
            {
                "mark": "circle",
                "encoding": {"x": {}, "y": {}, "color": {}, "size": {}},
            },
            {
                "mark": {"type": "line", "color": "red"},
                "transform": [{"regression": "field1", "on": "field2"}],
                "encoding": {"x": {}, "y": {}},
            },
        ],
    },
    "channels": ["x", "y", "size", "color", "column", "row"],
    "markCognitiveChannel": "position",
    "instantiate": _regression_instantiate,
    "properties": [
        {"key": "regressionMethod", "label": "Method", "type": "discrete",
         "options": [
             {"value": "linear", "label": "Linear"},
             {"value": "log", "label": "Logarithmic"},
             {"value": "exp", "label": "Exponential"},
             {"value": "pow", "label": "Power"},
             {"value": "quad", "label": "Quadratic"},
             {"value": "poly", "label": "Polynomial"},
         ],
         "defaultValue": "linear"},
        {"key": "polyOrder", "label": "Poly Order", "type": "continuous",
         "min": 2, "max": 10, "step": 1, "defaultValue": 3},
    ],
}


# ─── Ranged Dot Plot ────────────────────────────────────────────────────────

def _ranged_dot_instantiate(spec, ctx):
    encs = ctx["resolvedEncodings"]
    color = encs.get("color")
    rest = {ch: enc for ch, enc in encs.items() if ch != "color"}
    if "encoding" not in spec:
        spec["encoding"] = {}
    for ch, enc in rest.items():
        spec["encoding"][ch] = {**(spec["encoding"].get(ch) or {}), **enc}
    if color:
        layer1_enc = spec["layer"][1]["encoding"]
        layer1_enc["color"] = {**(layer1_enc.get("color") or {}), **color}

    encoding = spec.get("encoding") or {}
    y_enc = encoding.get("y")
    x_enc = encoding.get("x")
    if y_enc and y_enc.get("type") == "nominal":
        spec["layer"][0]["encoding"]["detail"] = copy.deepcopy(y_enc)
    elif x_enc and x_enc.get("type") == "nominal":
        spec["layer"][0]["encoding"]["detail"] = copy.deepcopy(x_enc)


ranged_dot_plot_def = {
    "chart": "Ranged Dot Plot",
    "template": {
        "encoding": {},
        "layer": [
            {"mark": "line", "encoding": {"detail": {}}},
            {"mark": {"type": "point", "filled": True}, "encoding": {"color": {}}},
        ],
    },
    "channels": ["x", "y", "color"],
    "markCognitiveChannel": "position",
    "instantiate": _ranged_dot_instantiate,
}


# ─── Boxplot ────────────────────────────────────────────────────────────────

def _boxplot_declare(cs, table, chart_properties):
    if not (cs.get("x") or {}).get("field") or not (cs.get("y") or {}).get("field"):
        return {}
    result = detect_banded_axis_force_discrete(cs, table, {"preferAxis": "x"})
    if not result:
        return {}
    return {
        "axisFlags": {result["axis"]: {"banded": True}},
        "resolvedTypes": result.get("resolvedTypes"),
        "paramOverrides": {"defaultBandSize": 28},
    }


def _boxplot_instantiate(spec, ctx):
    default_build_encodings(spec, ctx["resolvedEncodings"])
    layout = ctx.get("layout") or {}
    x_count = layout.get("xNominalCount", 0) or 0
    y_count = layout.get("yNominalCount", 0) or 0
    if x_count > 0 or y_count > 0:
        box_step = layout.get("xStep") if x_count > 0 else layout.get("yStep")
        if box_step is not None:
            from ...core import js_round
            box_size = max(4, js_round(box_step * 0.7))
            spec["mark"] = set_mark_prop(spec.get("mark"), "size", box_size)


boxplot_def = {
    "chart": "Boxplot",
    "template": {"mark": "boxplot", "encoding": {}},
    "channels": ["x", "y", "color", "opacity", "column", "row"],
    "markCognitiveChannel": "position",
    "declareLayoutMode": _boxplot_declare,
    "instantiate": _boxplot_instantiate,
}

File: packages/flint-js/src/vegalite/templates/scatter.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { ChartTemplateDef, ChartPropertyDef } from '../../core/types';
import {
    defaultBuildEncodings, applyPointSizeScaling, setMarkProp,
    detectBandedAxisForceDiscrete,
} from './utils';
import { makeCartesianPivot } from '../../core/pivot';

// Fraction of the band/lane step a boxplot box should occupy. An ungrouped box
// fills most of its category band; a grouped (dodged) box fills most of its
// per-subgroup lane. The remainder becomes the gap between adjacent boxes.
const BOXPLOT_BAND_FILL = 0.7;
const GROUPED_BOXPLOT_LANE_FILL = 0.85;
// Vega-Lite's default discrete position band scale reserves ~20% of each step as
// inter-band padding, so only ~80% of the step is usable drawing width. Grouped
// box sizing must use this usable width when splitting a band into sub-lanes,
// otherwise the boxes overshoot their lane pitch and overlap within a group.
const USABLE_BAND_FRACTION = 0.8;

export const scatterPlotDef: ChartTemplateDef = {
    chart: "Scatter Plot",
    template: { mark: "circle", encoding: {} },
    channels: ["x", "y", "color", "size", "shape", "opacity", "column", "row"],
    markCognitiveChannel: 'position',
    instantiate: (spec, ctx) => {
        defaultBuildEncodings(spec, ctx.resolvedEncodings);
        // A `shape` encoding only renders distinct glyphs on the `point` mark;
        // `circle` ignores it. Promote the mark when shape is in play.
        if (spec.encoding?.shape?.field) {
            spec.mark = setMarkProp(spec.mark, 'type', 'point');
        }
        applyPointSizeScaling(spec, ctx.table, ctx.canvasSize?.width, ctx.canvasSize?.height);
        const config = ctx.chartProperties;
        if (config?.opacity !== undefined && config.opacity < 1) {
            spec.mark = setMarkProp(spec.mark, 'opacity', config.opacity);
        }
    },
    properties: [
        { key: "opacity", label: "Opacity", type: "continuous", min: 0.1, max: 1, step: 0.05, defaultValue: 1 },
    ] as ChartPropertyDef[],
    pivot: makeCartesianPivot({
        // Flip the axes (orientation) as its own generator.
        transpose: [['x', 'y']],
        // x/y/color/size are peer measure channels: reassign a measure field
        // between a precise axis and a demoted color/size channel. Profile typing
        // prunes anything touching a discrete series; aux↔aux (color↔size) and
        // x↔y (a transpose) are not offered here.
        permute: [['x', 'y', 'color', 'size']],
        // Route the discrete grouping field across color / facet channels so a
        // grouped scatter and a faceted scatter are states of one another.
        shift: ['color', 'group', 'column', 'row'],
        // Chart-type transition: the discrete series field (wherever it sits —
        // color, column or row) moves onto the `x` category axis, re-rendering
        // the cloud as a Strip/Jitter plot. The displaced quantitative x spills
        // to a `color` gradient. Offered whenever a discrete series exists.
        transitions: [
            {
                to: 'Strip Plot',
                label: 'Jitter',
                route: { from: 'series', to: 'x', mode: 'swap', spill: 'color' },
            },
        ],
    }),
};

export const regressionDef: ChartTemplateDef = {
    chart: "Regression",
    template: {
        layer: [
            {
                mark: "circle",
                encoding: { x: {}, y: {}, color: {}, size: {} },
            },
            {
                mark: { type: "line", color: "red" },
                transform: [{ regression: "field1", on: "field2" }],
                encoding: { x: {}, y: {} },
            },
        ],
    },
    channels: ["x", "y", "size", "color", "column", "row"],
    markCognitiveChannel: 'position',
    instantiate: (spec, ctx) => {
        const { x, y, color, size, column, row } = ctx.resolvedEncodings;
        const config = ctx.chartProperties;
        // x & y → both layers + transform field names
        if (x) {
            spec.layer[0].encoding.x = { ...spec.layer[0].encoding.x, ...x };
            spec.layer[1].encoding.x = { ...spec.layer[1].encoding.x, ...x };
            if (x.field) spec.layer[1].transform[0].on = x.field;
        }
        if (y) {
            spec.layer[0].encoding.y = { ...spec.layer[0].encoding.y, ...y };
            spec.layer[1].encoding.y = { ...spec.layer[1].encoding.y, ...y };
            if (y.field) spec.layer[1].transform[0].regression = y.field;
        }
        // Regression method (default: linear)
        const method = config?.regressionMethod;
        if (method && method !== 'linear') {
            spec.layer[1].transform[0].method = method;
            // For polynomial, allow configurable order
            if (method === 'poly') {
                const order = config?.polyOrder ?? 3;
                spec.layer[1].transform[0].order = order;
            }
        }
        // color → scatter layer always; if present, also group regression by color field
        if (color) {
            spec.layer[0].encoding.color = { ...spec.layer[0].encoding.color, ...color };
            if (color.field) {
                // Group regression by color field so each class gets its own trend line
                spec.layer[1].transform[0].groupby = [color.field];
                // Pass color encoding to regression layer so lines match scatter colors
                spec.layer[1].encoding.color = { ...color };
                // Remove the hardcoded red so Vega-Lite uses the shared color scale
                spec.layer[1].mark = { type: "line" };
            }
        }
        if (size) spec.layer[0].encoding.size = { ...spec.layer[0].encoding.size, ...size };
        // facets → top-level encoding
        if (!spec.encoding) spec.encoding = {};
        if (column) spec.encoding.column = column;
        if (row) spec.encoding.row = row;
    },
    properties: [
        {
            key: "regressionMethod", label: "Method", type: "discrete",
            options: [
                { value: "linear", label: "Linear" },
                { value: "log",    label: "Logarithmic" },
                { value: "exp",    label: "Exponential" },
                { value: "pow",    label: "Power" },
                { value: "quad",   label: "Quadratic" },
                { value: "poly",   label: "Polynomial" },
            ],
            defaultValue: "linear",
        },
        {
            key: "polyOrder", label: "Poly Order", type: "continuous",
            min: 2, max: 10, step: 1, defaultValue: 3,
        },
    ] as ChartPropertyDef[],
};

export const rangedDotPlotDef: ChartTemplateDef = {
    chart: "Ranged Dot Plot",
    template: {
        encoding: {},
        layer: [
            { mark: "line", encoding: { detail: {} } },
            { mark: { type: "point", filled: true }, encoding: { color: {} } },
        ],
    },
    channels: ["x", "y", "color"],
    markCognitiveChannel: 'position',
    instantiate: (spec, ctx) => {
        const { color, ...rest } = ctx.resolvedEncodings;
        if (!spec.encoding) spec.encoding = {};
        for (const [ch, enc] of Object.entries(rest)) {
            spec.encoding[ch] = { ...(spec.encoding[ch] || {}), ...enc };
        }
        if (color) {
            spec.layer[1].encoding.color = { ...(spec.layer[1].encoding.color || {}), ...color };
        }

        // Copy nominal axis into detail encoding for line layer
        if (spec.encoding.y?.type === "nominal") {
            spec.layer[0].encoding.detail = JSON.parse(JSON.stringify(spec.encoding.y));
        } else if (spec.encoding.x?.type === "nominal") {
            spec.layer[0].encoding.detail = JSON.parse(JSON.stringify(spec.encoding.x));
        }
    },
};

export const boxplotDef: ChartTemplateDef = {
    chart: "Boxplot",
    template: { mark: "boxplot", encoding: {} },
    channels: ["x", "y", "color", "opacity", "column", "row"],
    markCognitiveChannel: 'position',
    declareLayoutMode: (cs, table) => {
        if (!cs.x?.field || !cs.y?.field) return {};
        const result = detectBandedAxisForceDiscrete(cs, table, { preferAxis: 'x' });
        if (!result) return {};
        return {
            axisFlags: { [result.axis]: { banded: true } },
            resolvedTypes: result.resolvedTypes,
            paramOverrides: { defaultBandSize: 28 },  // box+whisker needs wider bands
            colorActsAsGroup: true,  // dodge-by-color → budget band per category, shrink lanes
        };
    },
    instantiate: (spec, ctx) => {
        defaultBuildEncodings(spec, ctx.resolvedEncodings);

        const layout = ctx.layout;
        const hasDiscreteX = layout.xNominalCount > 0;
        const hasDiscreteAxis = hasDiscreteX || layout.yNominalCount > 0;

        // Grouped boxplots: a color field subdividing a categorical axis must
        // dodge the boxes side-by-side (xOffset/yOffset), not overlay them at the
        // same position — overlaid boxes hide whichever is drawn underneath.
        // Vega-Lite needs an explicit offset encoding to lay grouped boxes out.
        const colorEnc = spec.encoding?.color;
        let subgroups = 1;
        if (colorEnc?.field && hasDiscreteAxis && !spec.encoding.xOffset && !spec.encoding.yOffset) {
            const offsetChannel = hasDiscreteX ? 'xOffset' : 'yOffset';
            const offsetEnc: Record<string, unknown> = { field: colorEnc.field, type: 'nominal' };
            if (colorEnc.sort !== undefined) offsetEnc.sort = colorEnc.sort;
            spec.encoding[offsetChannel] = offsetEnc;
            const colorField = ctx.channelSemantics?.color?.field;
            if (colorField) {
                subgroups = Math.max(1, new Set(ctx.table.map((r) => r[colorField])).size);
            }
        }

        // Scale box width to the step size of the discrete axis. With
        // colorActsAsGroup, computeLayout sizes the axis per *category band*
        // (xStepUnit 'group') and Vega-Lite gets `width:{step, for:'position'}`,
        // so each band is subdivided into `subgroups` sub-lanes. Vega-Lite's
        // position band scale reserves ~20% of every step as inter-band padding,
        // so only ~80% of the step is actual drawing width — the per-subgroup
        // pitch is `step * USABLE_BAND_FRACTION / subgroups`. Sizing a box to the
        // raw `step / subgroups` overshoots that pitch and makes adjacent boxes
        // in a group overlap; account for the padding, then fill most of the lane
        // so boxes shrink (and stay separated) as subgroups grow.
        if (hasDiscreteAxis) {
            const boxStep = hasDiscreteX ? layout.xStep : layout.yStep;
            if (subgroups > 1) {
                const lanePitch = (boxStep * USABLE_BAND_FRACTION) / subgroups;
                const boxSize = Math.max(2, Math.round(lanePitch * GROUPED_BOXPLOT_LANE_FILL));
                spec.mark = setMarkProp(spec.mark, 'size', boxSize);
            } else {
                const boxSize = Math.max(4, Math.round(boxStep * BOXPLOT_BAND_FILL));
                spec.mark = setMarkProp(spec.mark, 'size', boxSize);
            }
        }
    },
};
